### ISSUE TITLE
How to implement custom widgets the Angular way?

### ISSUE THREAD

**Issue by halexiev-hedgeserv** [surveyjs/widgets#35]
Hi guys,

We had to switch to Angular from Aurelia and our custom widgets based on Bootstrap & jQuery stopped working. You can see the code below - basically we are overriding the default radiogroup and checkbox type of questions so we can have same look and feel across browsers as Bootstrap 4's widgets (i.e. 'custom-control' & 'custom-control-input') are SVG based and therefore consistent in all browsers.

Currently we get the following error:

-----------------------------------------------------------------------------------------------------
```
core.js:1448 ERROR Error: Unable to process binding "template: function (){return { name:question.koElementType,data:question,afterRender:$parent.koElementAfterRender} }"
Message: Unable to process binding "template: function (){return { name:'survey-rows',data:$data} }"
Message: Unable to process binding "foreach: function (){return { data:koRows,as:'row'} }"
Message: Unable to process binding "if: function (){return row.koVisible }"
Message: Unable to process binding "foreach: function (){return { data:row.koElements,as:'question',afterRender:row.koAfterRender} }"
Message: Unable to process binding "if: function (){return question.koVisible }"
Message: Unable to process binding "template: function (){return { name:question.koElementType,data:question,afterRender:$parent.koElementAfterRender} }"
Message: You cannot apply bindings multiple times to the same element.
    at applyBindingsToNodeInternal (survey.angular.js:5054)
    at applyBindingsToNodeAndDescendantsInternal (survey.angular.js:5000)
    at applyBindingsToDescendantsInternal (survey.angular.js:4982)
    at applyBindingsToNodeAndDescendantsInternal (survey.angular.js:5009)
    at applyBindingsToDescendantsInternal (survey.angular.js:4982)
    at applyBindingsToNodeAndDescendantsInternal (survey.angular.js:5009)
    at Object.ko.applyBindings (survey.angular.js:5211)
    at eval (survey.angular.js:7069)
    at invokeForEachNodeInContinuousRange (survey.angular.js:7020)
    at activateBindingsOnContinuousNodeArray (survey.angular.js:7067)
```
-----------------------------------------------------------------------------------------------------

**The question is - how would you do same thing in Angular terms and components?**

This is really important for us so any feedback and/or suggestions are very much appreciated.

```
  ngOnInit() {
    ...
    this.configureCheckEnabledWidget('checkbox', {'checkbox': true, 'boolean': true}, 'custom-checkbox');
    this.configureCheckEnabledWidget('radiogroup', {'radiogroup': true}, 'custom-radio');
  }

  private configureCheckEnabledWidget(type, fitTypes, cssClass) {
    let widget = {
      className: 'bootstrap-' + type,
      name: type,
      widgetIsLoaded: () => true,
      isFit: (question) => fitTypes[question.getType()],
      isDefaultRender: true,
      afterRender: (question, element) => {
        let $element = $(element);
        let $parent = $element.parent();
        $element.detach().find('label').each((index, label) => {
          let text = 'choices' in question ? question.choices[index].text : question.fullTitle;
          $(label).addClass('custom-control').addClass(cssClass)
                  .find('input').addClass('custom-control-input').end()
                  .children('span').remove().end()
                  .append(`<span class="custom-control-indicator"></span><span class="custom-control-description">${text}</span>`);
        }).end().appendTo($parent);
      }
    };

    Survey.CustomWidgetCollection.Instance.addCustomWidget(widget, 'type');
  }
```

Thank you,
Hristo

**Comment by tsv2013** [comment]
I think if you can try remove `detach` and `attachTo` for $element, this can help:

```JS
$element.find('label').each((index, label) => {
          let text = 'choices' in question ? question.choices[index].text : question.fullTitle;
          $(label).addClass('custom-control').addClass(cssClass)
                  .find('input').addClass('custom-control-input').end()
                  .children('span').remove().end()
                  .append(`<span class="custom-control-indicator"></span><span class="custom-control-description">${text}</span>`);
        }).end();
```

If not, could you provide us with a plunker or any other sandbox sample?

**Comment by halexiev-hedgeserv** [comment]
@tsv2013 Thank you, this helped in resolving the exception, but the widget is still not rendered properly.

Would it be possible to provide a clear example how to create a custom widget for Angular 5 using Bootstrap 4's css classes for checkboxes and radiogroups shown here:

https://getbootstrap.com/docs/4.0/components/forms/#checkboxes-and-radios-1

Basically in order to create a Bootstrap 4 checkbox you'd have the following html structure and css classes applied to it:

```
<div class="custom-control custom-checkbox">
  <input type="checkbox" class="custom-control-input" id="customCheck1">
  <label class="custom-control-label" for="customCheck1">Check this custom checkbox</label>
</div>
```

So how do you create such a custom checkbox/radiogroup widget in Surveyjs for Angular 5+?

Thank you,
Hristo

**Comment by tsv2013** [comment]
Glad to help you to resolve your first issue.

As for Angular 5 - our ordinary custom widgets (from this repo) are working with Angular 5 pretty well, you can check the https://github.com/surveyjs/surveyjs_angular_cli repo I've recently upgraded to the Angular 5.

If BS4 requires custom markup in your case, you can get the https://github.com/surveyjs/widgets/blob/master/src/pretty-checkbox.js as staring point I think

**Comment by halexiev-hedgeserv** [comment]
@tsv2013 I tried the pretty-checkbox route, but for some reason the "renderAs" property does not apply - is there some special configuration to make it work?

I tried the following:
Installed the pretty-checkbox npm package.

Added the following import in the global styles.scss:
`@import '~pretty-checkbox/src/pretty-checkbox.scss';`

Added the following In the surveyjs.component.ts:
```
import * as Survey from 'survey-angular';
import * as widgets from 'surveyjs-widgets/surveyjs-widgets.js'
widgets['prettycheckbox'](Survey);
```

Finally modified one of the questions in the json like this:
```
    {
      "elements": [
        {
          "type": "checkbox",
          "renderAs": "prettycheckbox",
          "name": "SidePocketElection",
          "title": "Election Not to Participate in Illiquid Sub-Capital Accounts. Check the box below if the Subscriber elects not to participate in Illiquid Sub-Capital Accounts, as described in Section 5.5(b)(i)(C) of the Agreement and in the Offering Circular. By checking this box, the Subscriber acknowledges that it will not be entitled to participate in the Master Fund’s investments in Illiquid Instruments.",
          "isRequired": true,
          "choices": [
            {
              "value": "SPLiq_1",
              "text": "Election Not to Participate in Illiquid Sub-Capital Accounts."
            }
          ]
        }
      ],
      "name": "SidePocketElectionSection",
      "title": "SidePocket Election",
      "visible": false,
      "visibleIf": "{InvestorType}='individual' or {InvestorType}='entity'"
    }

```

Is this a bug - any suggestions?

Thank you,
Hristo

**Comment by halexiev-hedgeserv** [comment]
Hey guys,

Apologies for the spam, but I am really having hard time with this issue above ^^^. I still cannot figure out why surveyjs does not pick up the 'renderAs' property ... your examples of Angular2 with pretty-checkbox are based on the purely js version, so could it be that in the angular cli case something's working a bit differently? Can you point to working cli based example?

Thank you,
Hristo

**Comment by tsv2013** [comment]
We've fixed "pretty-checkbox library was not found for bundled CSS" issue and updated the https://github.com/surveyjs/surveyjs_angular_cli repo (commit https://github.com/surveyjs/surveyjs_angular_cli/commit/c671bf13b10aebde549ca838541b43bfb83692cb)
Fix will be available in the next update

**Comment by tsv2013** [comment]
I've created a separate issue - https://github.com/surveyjs/widgets/issues/38
It's hard to find in this thread
